### PR TITLE
CP-46151: Productise the observer.py

### PR DESCRIPTION
**Pull request by snwoods**
Instead of being a noop, it will now automatically instrument the python
code given to it and create spans accordingly. The trace created will
link to the xapi code that called it. It will export to all endpoints
that it is given, which it reads from the observer.conf files created by
xapi.

This is built on top of https://github.com/xapi-project/xen-api/pull/5435 so please ignore the first two commits

For reviewers, by @bernhardkaindl:
- If you want to see the code (since Codecov adds annotations - it is not covered by a test at all), you've to press "a" to toggle the Codecov annotations off.
  - I submitted #5455 to configure Codecov so you can see the code without togging the annotations off.
  - It would be good to have some unit-test of course too, but this needs refactoring the code to use more functions before tests can be written for it. It should be refactored, formatted using `black` and `isort` into beautiful intelligible code, use `f"{var1}-{var2}"`-strings, not `var1 + "/" var2`, ethc, with comments and descriptive module and function `docstrings`.

### CODE REVIEW

**Comment by snwoods** [comment]
Here you can see an example trace with sm connecting to xapi

**Comment by bernhardkaindl** [comment]
@snwoods Hello, 

First of all: I like that it works and the screenshot, looks great!

But the code it is not beautiful intelligible code, but that should be very easy to fix.

I don't like you tell you about PEP8 because I guess the `max-columns=79` that it puts forward may not even be liked by the stdlib devs, but that does not make it invalid as a whole:

I think it's very important that the indentation level of 4 stays consistent within the project.

At the start of Raymond Hettinger's talk- _Beyond PEP 8 -- Best practices for beautiful intelligible code - PyCon 2015_, he calls an indentation of 2 "a little atrocity", You may like to watch the first 5:13 minutes, he is a really funny guy! Also, if you watch it further you'll see other formatting that is in your format in some places: https://youtu.be/wf-BqAjZb8M

But seriously, today we have code autoformatters so we can let them do `__all__` of the formatting for consistency.

Before I tried black myself, I hated beging "my" code formatted by a tool.

But one gets over it: After all, black only tweaks the constant syntax tree, it's safe as it was developed by a developer at Facebook at more than 10 million lines of source. Nowadays it's used in every second on gigantic masses of Python, including the massive Dropbox Python project(look up their Python3 migrating report).

But specifically, don't even get distracted by formatting/line breaks, as black just takes care if the formatting, so it gets out of the way, no brain cell is involved with formatting anymore, more are available to focus on the code.

You may also see in the CI that  the pylint-based reviewdog code review action is overwhelmed by the indendation errors.

So you just run `pip install black && black scripts/observer.py`. Formatting is: Done.

That brings the code into a shape that pylint can work on better.

Maybe look here for how the PR would look with black: https://github.com/xenserver-next/xen-api/pull/15

For a start, maybe ignore the pylint `import-error` warnings, but they remind us:
For XS8, I found:
- python-opentelemetry.spec (but maybe it was not merged to master yet - I guess also wrapt is there?)
  - But I have not found opentelemetry-exporter-zipkin (maybe is this included?
  
But I tink you should adda note for when it is merged to add them into the Requires of the spec file?

I just paste some pylints:
- W4902: Using deprecated method getcallargs() (deprecated-method)
  - If you can't use the new replacement, add `# pylint: disable-next=deprecated-method` before it.

# Some you may disable on the top of the file, because of the decorator stuff...:
`# pylint: disable=no-value-for-parameter,unused-argument`
- W0613: Unused argument 'instance' (unused-argument)`, where it's needed for the decorator
- E1120: No value for argument 'kwargs' in function call (no-value-for-parameter)
 
But please fix this, maybe use `exc` in this (or the outer name):
- W0621: Redefining name 'e' from outer scope (line 32) (redefined-outer-name)
If the API is, good, it should not raise `Exception` but a subclass of it, catch the subclass insted:
- W0718: Catching too general exception Exception (broad-exception-caught)

Please goup all the stdlib imports on top, they do not need to be in the middle of the code.

And please also have a look at this short review:
https://app.codacy.com/gh/xenserver-next/xen-api/pullRequest?prid=13605604

Please check if you can reduce the complexy. At the very least factor this into a separate function,
but maybe, prefeably it should be it's own module as you (understandably) you do not want to import all the opentelemetry imports by default.

McCabe check says: `If 53 is too complex (39) (MC0001)`:
```py
# Do not do anything unless configuration files have been found
if configs:
    tracers = []
    ... lots of code, it's imports and does many things ...
```

Just to clarify - I think some documentation would be good to understand what it does,
maybe in scripts/docs/README-observer.md

File: scripts/observer.py
Comment (bernhardkaindl) [inline scripts/observer.py:97]: This overlong string lambda would be much nicer to read and understand this way:
```suggestion
    def bugtool_filenamer()
        """Compound function to return an rfc3339-compilant ndjson file name"""
        now = datetime.now(timezone.utc).isoformat()
        f"{trace_log_dir}/{service_name}-{host_uuid}-{tracestate}-{now}.ndjson"
```

**Review by bernhardkaindl — CHANGES_REQUESTED** [review]

**Comment by snwoods** [comment]
Hey @bernhardkaindl thank you so much for your very thorough review! Before seeing your comment I refactored observer.py a little to use functions and make it more testable, but I need to add a bit more coverage still. I'll now work on beautifying the code using the helpful tools you mentioned :)

**Comment by snwoods** [comment]
This should be in a good shape now: we have enough code coverage to pass the coverage check, I've reformatted and I've fixed all the pylint errors.

The only things outstanding is to install wrapt as a dependency into the github CI (this will get more code coverage as we can actually test the main wrapper function in span_of_tracers) and to do something about the python2 unittests. observer.py only works for python3 as the opentelemetry library it depends on only supports python3. Does this mean we should remove it from the python2 unittests check somehow?

File: ocaml/xapi/xapi_observer.ml
Comment (snwoods) [reply ocaml/xapi/xapi_observer.ml:263]: The issue is that the DB is not an authoritative source. As seen here, the default_attributes comes from outside the DB and is given on top of the user provided attributes in set_attributes. The other component's forwarders remember this information inside their TracerProvider object and use that as an authoritative source, not the DB. If they used the DB as an authoritative source, they would forget the default_attributes whenever a different parameter is set e.g. endpoints.

**Comment by snwoods** [comment]
I have added wrapt as a dependency, so the only thing outstanding is the python2 tests.
I have skipped the test for python2 by checking sys.version_info but it has negatively affected the python2.7 codecov, bringing it below the threshold...

File: ocaml/xapi/xapi_observer.ml
Comment (mg12) [reply ocaml/xapi/xapi_observer.ml:263]: keeping the code the way it is now seems like the correct thing to do. The config_of_observer ends up being the attributes of the DB (which are generic across all the components) + the default_attributes of the component. So the DB is the authoritative source for all the components, and the default_attributes is the authoritative source for a specific component.

**Comment by edwintorok** [comment]
> but it has negatively affected the python2.7 codecov, bringing it below the threshold...

Well thats good, we don't want to add any *new* Python2.x code if we can avoid it. 
If we have Python3-only new code then that code shouldn't be counted when python2 code coverage is calculated. @bernhardkaindl is there some way to tell the CI/codecov that a particular file is py3 only and ignore counting it for py2?

**Comment by snwoods** [comment]
This should be ready now. The python2 test coverage issue was resolved by moving observer.py (and all future python3-only scripts) to a new directory python3/ which is not included in the python2 pytest command.
I've tested this again on a host with the latest build:
image

**Comment by stephenchengCloud** [comment]
May need to add this in pyproject.toml to apply the pytype check to python3 scripts. @bernhardkaindl 
> diff --git a/pyproject.toml b/pyproject.toml
> index b65a36bb0..8a423c4da 100644
> --- a/pyproject.toml
> +++ b/pyproject.toml
> @@ -114,6 +114,9 @@ inputs = [
>      "scripts/yum-plugins",
>      "scripts/*.py",
>  
> +    # Python 3
> +    "python3/*.py",
> +
>      # To be added later,
>      # when converted to Python3-compatible syntax:
>      # "ocaml/message-switch/python",

**Review by bernhardkaindl — APPROVED** [review]
I approve!

I like this PR very much, and I'm happy that this might become a new start into a nice new epoch of Python code!

I had submitted some suggestions (I submitted them separately in https://github.com/snwoods/xen-api/pull/1) but these should not block the merge!

Very nice!

**Comment by snwoods** [comment]
I've updated the PR with Bernhard's suggestions. If you could give it a quick once over then I'll merge it :) I created branch private/stevenwo/CP-46151-pre-bernhard-changes so that you could easily compare what's been changed but github's compare doesn't seem up to the task...

**Review by bernhardkaindl — APPROVED** [review]
I reviewed the changes, ran the test locally too, CI has passed.

For use on Dom0, it will need the `opentelemetry` RPM packages added as `Requires:` to the `xapi.spec`

**Pull request merged**
